### DISCUSSION TITLE
Add `#[non_exhaustive]` to protocol event `enum`s

### DISCUSSION THREAD

**Discussion by dhuseby** [libp2p/rust-libp2p#4543] [Ideas]
While working on PR #4503 a [short discussion](https://github.com/libp2p/rust-libp2p/pull/4503#issuecomment-1726968553) came up about how adding an emitted event to the Kademlia protocol was a breaking change. I wondered out loud to @thomaseizinger if we should also add `#[non_exhaustive]` to the event `enum` as part of this breaking change so that in the future it won't be a breaking change to add more events. I brought the discussion here to surface it and build some consensus.

Proposal:
* Add `#[non_exhaustive]` to the event `enum` for protocols as a one-time breaking change.

Pros:
* Adding events to the protocol `enum` in the future won't trigger any compile warnings.
* Doesn't affect any code within the protocol crate is that attribute is ignored per the RFC.
* Changing an existing event does trigger compile warnings.
* Most downstream code already has a wildcard arm: `_ => {}`

Cons:
* All downstream dependent code that `match`es against the event `enum` must have a wildcard arm: `_ => {}`
* The only way to get notified of new events in the `enum` is to enable a—currently unstable—lint check: https://github.com/rust-lang/rust/issues/89554

Thoughts? @thomaseizinger mentioned that we had considered it before but I couldn't find any previous posts in theses discussions or on [discuss.libp2p.io](https://discuss.libp2p.io).

**Comment by thomaseizinger** [comment]
The more I think about it, the more I like this idea. I'll bring it up in today's maintainer call: https://github.com/libp2p/rust-libp2p/discussions/4556

**Comment by thomaseizinger** [comment]
Here is the full description of this feature: https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute

One thing that is nice and I didn't know is that it can also be applied to `enum` variants which allows you to add more fields to it without it being a breaking change.

As an example, if we would have `non_exhaustive` on `FromSwarm`, we could ship https://github.com/libp2p/rust-libp2p/pull/4371 as a non-breaking change which would be quite nice.

Similiarly, if all `Event`s would have `non_exhaustive` applied to their variants, we could have shipped https://github.com/libp2p/rust-libp2p/pull/4527 as a non-breaking change.

**Comment by thomaseizinger** [comment]
Perhaps a suitable policy would be to at least apply this to enums from `libp2p-core` and `libp2p-swarm`. Breaking changes to those have a cascading effect across the workspace and often hold us back in shipping certain features.

**Reply by dhuseby** [reply]
Don't forget that adding the attribute is itself a breaking change so we'll have to do it carefully and in a controlled way.

**Comment by mxinden** [comment]
I am in favor of the proposal. I think we should consider `#[non_exhaustive]` for multiple `pub` `enum`s across the code base. Thanks @dhuseby for the suggestion and thanks @thomaseizinger for bringing it up in the last open maintainers call.

Another con: in case we ever have a semantically breaking change, we can no longer force a user to handle the case of the change by adding a new enum variant. Though off the top of my head, I can not think of such a need in the past.

**Reply by thomaseizinger** [reply]
Nice! Should we do that as part of the upcoming breaking release already or shall we let this idea sit a bit more to figure out, which enums it is worth adding to?

I think adding it to _every_ enum (and its variants) might be a bit excessive, no? If we add it to `FromSwarm` and `ToSwarm`, we also need to figure out how we want to handle `libp2p-swarm-derive`. Because this generates code in user-land, `non_exhaustive` is going to force us to generate a `_` arm.

Let's say we add a new variant to `FromSwarm`. I think that changing `libp2p-swarm-derive` to generate code for this new variant is a breaking change of the macro crate. If we were to release it as a patch-release, a user could _technically_ only bump `libp2p-swarm-derive` and would run into a compile-error because it generates code that the version of `libp2p-swarm` they are depending on doesn't have. This is the classic: "We must depend on the latest version of the crate"-dilemma, except that `libp2p-swarm-derive` does not have a cargo dependency to `libp2p-swarm`. Thus, I think we must not release such changes to `libp2p-swarm-derive` as patch releases.

I think what we have to do is release them as minor releases but bump the version in `libp2p-swarm` in a patch release, i.e. assuming `libp2p-swarm` bumps from `0.43.0` to `0.43.1` and adds a new variant, then the accompanying `libp2p-swarm-derive` version bump has to be `0.33.0` to `0.34.0`. This prevents the user from running a `cargo update -p libp2p-swarm-derive`. At the same time, I believe we can treat a minor version bump of the macro dependency as a non-breaking change in `libp2p-swarm` because the macro doesn't actually expose any symbols that a user could depend on.

This is really subtle :(

**Reply by dhuseby** [reply]
Next steps are to decide on which enums get this and which ones don't. @thomaseizinger you said `FromSwarm` and `ToSwarm`? What else?

**Reply by thomaseizinger** [reply]
Probably:

- `SwarmEvent`
- `ConnectionHandlerEvent`
- `ConnectionEvent`

**Reply by thomaseizinger** [reply]
> I think we must not release such changes to `libp2p-swarm-derive` as patch releases.

Any ideas on how we can prevent that? I can already see us running into this problem because it is super subtle and we just forget about that.

Should we add a comment next to `version = ` in `libp2p-swarm-derive`'s `Cargo.toml` that says something like "READ THIS BEFORE YOU BUMP THE VERSION" and link to a document explaining how to do the bump?

**Reply by thomaseizinger** [reply]
So, actually, I think the best solution to this is to always _pin_ the `libp2p-swarm-derive` dependency in `libp2p-swarm`. That way, it can never get out of sync, regardless of which version upgrades we are making.

**Reply by mxinden** [reply]
> So, actually, I think the best solution to this is to always _pin_ the `libp2p-swarm-derive` dependency in `libp2p-swarm`. That way, it can never get out of sync, regardless of which version upgrades we are making.

Sounds like the best option to me. I can not see a downside to this, given that we control both `libp2p-swarm` and `libp2p-swarm-derive`.

**Comment by dhuseby** [comment]
draft PR #4581 submitted
